gateway/server.go

```diff
@@ -131,25 +133,25 @@ func main() {
 
 	if credentials != nil {
```

Comment:
It seems odd for this to be based on the credentials, the credentials shouldn't be used in the gateway anymore and are specific to only one type of auth.  It seems like the check for if there should be auth should be handled by the value for `config.AuthProxyURL` and the parsing of the credentials should be removed

Reply:
Please see the diagram in the attached issue.

The secret is used for core-service commutation. The plugin is used for user-auth to the gateway.

Reply:
There seems to only be one auth configuration, via the `MakeExternalAuthHandler`. How can the internal service communication use basic-auth if the auth proxy points at an external authentication that is not basic auth?  This works right now, because the only external proxy is the basic auth verification but it seems like it would break as soon as you tried to validate with any other external auth proxy.

Reply:
You are correct. I think the upstreamProxy needs to have the credentials for the upstream calls to authenticated endpoints.

Reply:
Function proxy would remain as it is:

```
	faasHandlers.Proxy = handlers.MakeForwardingProxyHandler(reverseProxy, functionNotifiers, functionURLResolver, functionURLTransformer)
```

List for instance, would take a new function override so that it could provide relevant headers for calls.. 

```
func applyServiceAuth(r *http.Request) {
 r.Header.SetBasicAuth(user, pass)
}

	faasHandlers.ListFunctions = handlers.MakeForwardingProxyHandler(reverseProxy, forwardingNotifiers, urlResolver, nilURLTransformer, applyServiceAuth)
```

Reply:
Please see: https://github.com/openfaas/faas/pull/1214